content/guides/use-case/tensorflowjs.md

```diff
@@ -0,0 +1,584 @@
+---
+description: Learn how to deploy pre-trained models in a TensorFlow.js web applications to perform face detection.
+keywords: tensorflow.js, machine learning, ml, mediapipe
+title: Face detection with TensorFlow.js
+toc_max: 2
+---
+
+This guide introduces the seamless integration of TensorFlow.js with Docker to
+perform face detection. In this guide, you'll explore:
+
+- How to run a containerized TensorFlow.js application with Docker
+- The code for a face detection application using TensorFlow.js
+- How to create and modify the Docker assets for a TensorFlow.js application
+
+> **Acknowledgment**
+>
+> Docker would like to thank [Harsh Manvar](https://github.com/harsh4870) for
+> his contribution to the this guide.
+
+## Prerequisites
+
+* You have installed the latest version of [Docker Desktop](../../../get-docker.md).
+* You have a [Git client](https://git-scm.com/downloads). The examples in this guide use a command-line based Git client, but you can use any client.
+
+## What is TensorFlow.js?
+
+[TensorFlow.js](https://www.tensorflow.org/js) is an open-source JavaScript
+library for machine learning that enables you to train and deploy ML models in
+the browser or on Node.js. It supports creating new models from scratch or using
+pre-trained models, facilitating a wide range of ML applications directly in web
+environments. TensorFlow.js offers efficient computation, making sophisticated
+ML tasks accessible to web developers without deep ML expertise.
+
+## Why Use TensorFlow.js and Docker together?
+
+- Environment consistency and simplified deployment: Docker packages
+  TensorFlow.js applications and their dependencies into containers, ensuring
+  consistent runs across all environments and simplifying deployment.
+- Efficient development and easy scaling: Docker enhances development efficiency
+  with features like hot reloading and facilitates easy scaling of -
+  TensorFlow.js applications using orchestration tools like Kubernetes.
+- Isolation and enhanced security: Docker isolates TensorFlow.js applications in
+  secure environments, minimizing conflicts and security vulnerabilities while
+  running applications with limited permissions.
+
+
+## Get and run the sample application
+
+In a terminal, clone the sample application using the following command.
+
+```console
+$ git clone https://github.com/docker/face-detection-tensorjs
+```
+
+After cloning the application, you'll notice the application has a `Dockerfile`
+and a `compose.yaml` file. These Docker assets let you build and run the
+application locally with nothing more than Docker.
+
+To run the application, run the following command in a terminal
+inside the `face-detection-tensorjs` directory.
+
+```console
+$ docker compose up --build
+```
```

Comment:
Building doesn't seem to work with the given Dockerfile in https://github.com/craig-osterhout/face-detection-tensorjs

It's missing python and make at least. If you swap out the alpine version of the node image to something Debian-based, like `node:20-bookworm`, and remove the `apk add`s, it seems to build fine.

Reply:
It works on my Windows machine 😅
Thanks for testing. I'll dig into this more.

Reply:
Updated to the debian-based image.

Reply:
🤔 I'm getting a build error with node 21

```
12.36 [5/5] Building fresh packages...
12.79 info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
12.79 error /usr/src/app/node_modules/deasync: Command failed.
12.79 Exit code: 1
12.79 Command: node ./build.js
12.79 Arguments: 
12.79 Directory: /usr/src/app/node_modules/deasync
12.79 Output:
12.79 node:events:493
12.79       throw er; // Unhandled 'error' event
12.79       ^
12.79 
```

Node 20 doesn't error out. I have no idea why. It looks like the original reference project was using node 18 in the slim version. Maybe there was a breaking change between node 20 and 21. Unless you want to dig into the JS code here I guess using node 20 is fine. (It's LTS anyway: https://nodejs.org/en/about/previous-releases)

Reply:
Resolving this. The sample app has been updated to no longer build a Node.js app.